### PR TITLE
[Merged by Bors] - feat: Add support for building documentation via facets

### PR DESCRIPTION
**Pull request by hargoniX**
This adds supports for building the documentation via a Lake facet as described in the README. The `meta if` is in place so that mathlib can still update a nightly version and have doc-gen breaking without its own build breaking in order for doc-gen not to becoome the critical path of a nightly update. Furthermore it also helps compile times as barely any users wants to actually use it.

The facet itself also has a huge performance advantage, on my local machine the old doc-gen4 from master with the unified (non parallel) build approach took 8-9 min to run. The new one that uses the full power of being parallelized by Lake on a per module basis only takes 2 minutes.

Furthermore once this PR is merged the CI of doc-gen4 should run through and deploy the first version that has (at least partial) support for linking back from the LeanInk rendered code into doc-gen pages.

### CODE REVIEW

File: lakefile.lean [hunk 1]
```diff
@@ -11,3 +11,6 @@ lean_lib Mathlib
 lean_exe runLinter where
   root := `scripts.runLinter
   supportInterpreter := true
+
+meta if get_config? env = some "dev" then -- dev is so not everyone has to build it
```
Comment (gebner) [inline lakefile.lean:15]:
```suggestion
meta if get_config? doc = some "on" then -- do not download and build doc-gen4 by default
```
The `env` option doesn't mesh well with `lake update`, at least if we ever add a second env.  For `lake update`, all dependencies need to be present.  This isn't possible with `lake -Kenv=dev -Kenv=test update`, but only with separate options `lake -Kdoc=on -Ktest=on update`.

File: lean_packages/manifest.json [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+   "rev": "439303af06465824588a486f5f9c023ca69979f3",
+   "name": "leanInk"},
+  {"url": "https://github.com/leanprover/doc-gen4",
+   "rev": "a2345380e861332fabecc61591cb274e61163a97",
+   "name": "doc-gen4",
+   "inputRev": "main"},
+  {"url": "https://github.com/xubaiw/Unicode.lean",
+   "rev": "25b32ef69fd6d8ae34300d51dc0b9403ffb1fd29",
+   "name": "Unicode"},
+  {"url": "https://github.com/leanprover-community/mathlib4.git",
```
Comment (gebner) [inline lean_packages/manifest.json:16]: Why does doc-gen4 depend on mathlib4?

**Review by gebner — COMMENTED** [review]
Okay, this is adding a lot of dependencies.  And we're already running into the first issue with Lean version incompatibilities: https://github.com/leanprover/doc-gen4/pull/80

Do you have any idea on how responsive the maintainers are?

What is your preferred workflow to update Lean versions?  This will be a bit of a pain when doc-gen4 CI depends on mathlib4.  We'll have to update them in sync.

File: lean_packages/manifest.json (hunk 2 above)
Comment (hargoniX) [reply lean_packages/manifest.json:16]: doc-gen itself does not depend on mathlib for sure. If I grep for mathlib in a doc-gen build I get:
```
λ grep -r -i "mathlib" . | grep lakefile.lean
grep: ./lean_packages/mathlib/.git/index: binary file matches
grep: ./lean_packages/mathlib/.git/objects/pack/pack-0a95fa21965d4516ca298f75aa1f0b2a838b9b47.pack: binary file matches
grep: ./lean_packages/mathlib/.git/objects/pack/pack-f021064ed325132390895a54799a5c0197b0ef8a.pack: binary file matches
grep: ./lean_packages/mathlib/build/lib/Mathlib/Util/IncludeStr.olean: binary file matches
grep: ./lean_packages/mathlib/build/ir/Mathlib/Util/IncludeStr.o: binary file matches
./lean_packages/mathlib/lakefile.lean:package mathlib
./lean_packages/mathlib/lakefile.lean:lean_lib Mathlib
grep: ./lean_packages/lake/.git/objects/pack/pack-aa696e402774de9cdf8ccf888450b9d59cfe48af.pack: binary file matches
grep: ./lean_packages/lake/build/lib/Lake/CLI/Help.olean: binary file matches
grep: ./lean_packages/lake/build/lib/Lake/CLI/Init.olean: binary file matches
grep: ./lean_packages/lake/build/ir/Lake/CLI/Help.o: binary file matches
grep: ./lean_packages/lake/build/ir/Lake/CLI/Init.o: binary file matches
grep: ./build/lib/DocGen4/Output/Template.olean: binary file matches
./lean_packages/leanInk/test/dep/lakefile.lean:require mathlib from git "https://github.com/leanprover-community/mathlib4.git"
grep: ./build/bin/doc-gen4: binary file matches
grep: ./build/ir/DocGen4/Output/Template.o: binary file matches
```
so it would seem to me that `LeanInk` is the culprit here but also only with a test? Either way the `build` directory of mathlib4 inside of `lean_packages` of a doc-gen4 build is empty so it is definitely not a hard requirement for building the binary.

**Comment by hargoniX** [comment]
The version incompatibilities between doc-gen and mathlib4 have existed for much longer than this already due to the doc-gen CI trying to build mathlib4 docs :D. Usually mathlib4 is the reason that I wait or push for the next version update on master so there isn't really a change with this, I've more or less come to terms with my dependency on mathlib4 already.

The idea of adding the optional flag was to make it so that this dependency is not bidirectional, i.e. mathlib4 can keep updating as they like without breaking their main build and doc-gen has to move with them at some point in time (or alternatively has already moved further and mathlib4 has not).

For making the CI on doc-gen's side work I've been thinking about substituting the `require` line to a local file system dependency inside of mathlib4s `lakefile.lean` inside of my CI though I do not know how brittle this might be.

Maybe @tydeu has an idea on how one might nicely do such a thing with lake?

File: lean_packages/manifest.json (hunk 2 above)
Comment (gebner) [reply lean_packages/manifest.json:16]: I think this is a bug with `lake`.  Apparently it won't delete old manifest entries.  If I do `rm lean_packages/manifest.json && lake update` in `doc-gen4`, then the `mathlib4` entry goes away.

**Comment by gebner** [comment]
> mathlib4 can keep updating as they like without breaking their main build

Ah, then I understood this the wrong way.  I thought the plan was to enable CI for docs in both mathlib4 and doc-gen4.

File: README.md [hunk 3]
```diff
@@ -32,3 +32,9 @@ and the "real" port hasn't started yet
   ```
   find Mathlib -name "*.lean" | env LC_ALL=C sort | sed 's/\.lean//;s,/,.,g;s/^/import /' > Mathlib.lean
   ```
+## Building HTML documentation
+Building HTML documentation locally is straightforward:
+```
+lake -Kenv=dev build Mathlib:docs
```
Comment (gebner) [inline README.md:38]:
```suggestion
lake -Kdoc=on build Mathlib:docs
```

**Comment by gebner** [comment]
bors merge
